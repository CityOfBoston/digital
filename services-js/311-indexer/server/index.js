// @flow
/* eslint no-console: 0 */

// Entrypoint for our server. Uses require so we can control import order
// and set up error reporting before getting the main server.js file going.

console.log('----- SERVER STARTUP -----');

require('dotenv').config();
const opbeat = require('opbeat/start');

// This monkeypatches the environment to add window and an XMLHttpRequest
// implementation.
require('cometd-nodejs-client').adapt();

// The above 'adapt' creates a window object because the cometd JS library
// assumes a browser environment. But, the RxJS library prefers "window" as its
// global object, and needs an Object defined on it.
//
// eslint-disable-next-line no-undef
Object.assign(window, {
  Object,
  setInterval,
  clearInterval,
});

const start = require('./server.js').default;

start({ opbeat }).catch(err => {
  console.log('----- SERVER FAILURE -----');
  console.error(err);

  // We use process.kill to trigger any cleanup actions. process.exit won't let
  // us do async cleanup through the node-cleanup library.
  process.kill(process.pid, 'SIGTERM');
});
