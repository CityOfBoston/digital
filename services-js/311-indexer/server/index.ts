/* eslint no-console: 0 */

// Entrypoint for our server. Uses require so we can control import order
// and set up error reporting before getting the main server.js file going.

console.log('----- SERVER STARTUP -----');

require('dotenv').config();

const Rollbar = require('rollbar');
const rollbar = new Rollbar({
  accessToken: process.env.ROLLBAR_ACCESS_TOKEN,
  captureUncaught: true,
  captureUnhandledRejections: true,
  payload: {
    environment: process.env.ROLLBAR_ENVIRONMENT || process.env.NODE_ENV,
  },
});

const HttpsProxyAgent = require('https-proxy-agent');

// This monkeypatches the environment to add window and an XMLHttpRequest
// implementation.
require('./vendor/cometd-nodejs-client').adapt(
  process.env.http_proxy
    ? () => new HttpsProxyAgent(process.env.http_proxy)
    : undefined
);

// The above 'adapt' creates a window object because the cometd JS library
// assumes a browser environment. But, the RxJS library prefers "window" as its
// global object, and needs an Object defined on it.
//
// eslint-disable-next-line no-undef
Object.assign(window, {
  Object,
  setInterval,
  clearInterval,
});

const start = require('./server.js').default;

start({ rollbar }).catch(err => {
  console.log('----- SERVER FAILURE -----');
  console.error(err);

  // We use process.kill to trigger any cleanup actions. process.exit won't let
  // us do async cleanup through the node-cleanup library.
  process.kill(process.pid, 'SIGTERM');
});
