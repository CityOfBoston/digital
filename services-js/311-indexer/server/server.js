// @flow
/* eslint no-console: 0 */

import cleanup from 'node-cleanup';
import Elasticsearch from './services/Elasticsearch';
import Salesforce from './services/Salesforce';

type Opbeat = $Exports<'opbeat'>;

type ServerArgs = {
  opbeat: Opbeat,
};

type Invoice = {|
  Description__c: string,
  Id: string,
  Status__c: string,
  Name: string,
|};

export default async function startServer(args: ServerArgs) {
  if (
    process.env.ELASTICSEARCH_URL &&
    process.env.ELASTICSEARCH_URL.endsWith('.amazonaws.com')
  ) {
    Elasticsearch.configureAws(process.env.AWS_REGION);
  }

  const elasticsearch = new Elasticsearch(
    process.env.ELASTICSEARCH_URL,
    process.env.ELASTICSEARCH_INDEX,
    args.opbeat
  );

  try {
    const info = await elasticsearch.info();
    console.log('Elasticsearch connected', info);
  } catch (err) {
    console.error('ERROR GETTING ELASTICSEARCH INFO', err);
  }

  const salesforce: Salesforce<Invoice> = new Salesforce(
    process.env.SALESFORCE_COMETD_URL,
    process.env.SALESFORCE_PUSH_TOPIC,
    process.env.SALESFORCE_CONSUMER_KEY,
    process.env.SALESFORCE_CONSUMER_SECRET,
    args.opbeat
  );

  salesforce.on('meta', (msg: Object) => {
    console.info(JSON.stringify(msg));
  });

  salesforce.on('error', (err: Error) => {
    // This may happen if your session expires. Rather than try to re-auth, just exit
    // and let the container agent restart us.
    console.error('Error in Salesforce connection. Exiting.', err);
    process.exit(1);
  });

  await salesforce.connect(
    process.env.SALESFORCE_OAUTH_URL,
    process.env.SALESFORCE_API_USERNAME,
    process.env.SALESFORCE_API_PASSWORD,
    process.env.SALESFORCE_API_SECURITY_TOKEN
  );

  const shutdown = async () => {
    await salesforce.disconnect();
  };

  cleanup(exitCode => {
    shutdown().then(
      () => {
        process.exit(exitCode || 0);
      },
      err => {
        console.log('CLEAN EXIT FAILED', err);
        process.exit(-1);
      }
    );

    cleanup.uninstall();
    return false;
  });
}
