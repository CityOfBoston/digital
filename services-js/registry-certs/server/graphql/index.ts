import fs from 'fs';
import path from 'path';
import { GraphQLScalarType, Kind } from 'graphql';
import { makeExecutableSchema } from 'graphql-tools';
import Stripe from 'stripe';
import Rollbar from 'rollbar';

import { Query, resolvers as queryResolvers } from './query';
import { Mutation, resolvers as mutationResolvers } from './mutation';
import { resolvers as deathResolvers } from './death-certificates';

import RegistryDb from '../services/RegistryDb';
import Emails from '../services/Emails';

// This file is built by the "generate-graphql-schema" script from
// the above interfaces.
const schemaGraphql = fs.readFileSync(
  path.resolve(
    // Normalize between dev and compiled file locations
    __dirname.replace('/registry-certs/build/', '/registry-certs/'),
    '../../graphql',
    'schema.graphql'
  ),
  'utf-8'
);

export interface Context {
  rollbar: Rollbar;
  registryDb: RegistryDb;
  stripe: Stripe;
  emails: Emails;
}

/** @graphql schema */
export interface Schema {
  query: Query;
  mutation: Mutation;
}

const dateResolvers = {
  Date: new GraphQLScalarType({
    name: 'Date',
    description:
      'JavaScript Date as an ISO string or number of millis since Epoch',
    parseValue(value: any): Date | null {
      if (typeof value === 'string' || typeof value === 'number') {
        return new Date(value);
      } else {
        return null;
      }
    },
    parseLiteral(ast): Date | null {
      if (ast.kind === Kind.STRING) {
        return new Date(ast.value); // ast value is always in string format
      } else if (ast.kind === Kind.INT) {
        return new Date(parseInt(ast.value, 10));
      } else {
        return null;
      }
    },
    serialize(value: Date) {
      return value.toISOString();
    },
  }),
};

export default makeExecutableSchema({
  typeDefs: [schemaGraphql],
  resolvers: {
    ...queryResolvers,
    ...mutationResolvers,
    ...deathResolvers,
    ...dateResolvers,
  } as any,
  allowUndefinedInResolve: false,
});
