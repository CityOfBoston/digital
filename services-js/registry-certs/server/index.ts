/* eslint no-console: 0 */

// Entrypoint for our server. Uses require so we can control import order
// and set up error reporting before getting the main server.js file going.

require('dotenv').config();

const Rollbar = require('rollbar');
const rollbar = new Rollbar({
  accessToken: process.env.ROLLBAR_ACCESS_TOKEN,
  captureUncaught: true,
  captureUnhandledRejections: true,
  payload: {
    environment: process.env.ROLLBAR_ENVIRONMENT || process.env.NODE_ENV,
  },
});

try {
  const start = require('./server').default;

  start({ rollbar }).catch(err => {
    console.error('Error starting server');
    console.error(err);
    rollbar.error(err, () => {
      process.exit(1);
    });
  });
} catch (e) {
  console.error(e);
  process.exit(-1);
}
