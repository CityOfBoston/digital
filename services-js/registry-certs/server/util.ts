import path from 'path';

/**
 * Absolute path to the root of our subpackage within the monorepo.
 */
export const PACKAGE_SRC_ROOT: string = path.resolve(
  // We normalize between this file being run from the source directory and
  // being run from the build directory (which is one level deeper).
  __dirname.replace('/registry-certs/build/', '/registry-certs/'),
  '../'
);

/**
 * The representation of a file upload when Hapiâ€™s multipart handling is set to
 * "annotated"
 */
export interface AnnotatedFilePart {
  filename: string;
  headers: Object;
  payload: Buffer;
}
