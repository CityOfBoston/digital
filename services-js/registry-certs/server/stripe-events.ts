/* eslint no-console: 0 */
import RegistryDb from './services/RegistryDb';
import Emails from './services/Emails';
import Stripe, { charges, webhooks } from 'stripe';

import { loadOrder } from './graphql/death-certificates';

import {
  FIXED_CC_SERVICE_FEE,
  PERCENTAGE_CC_SERVICE_FEE,
  SERVICE_FEE_URI,
} from '../lib/costs';

interface Dependencies {
  emails: Emails;
  stripe: Stripe;
  registryDb: RegistryDb;
}

// We handle Stripe’s "charge.succeeded" event to complete the order process:
//   - Mark the order as "paid" in the backend so that fulfillment can complete
//   - Send a receipt email
//
// We do this from Stripe’s webhook to take advantage of their reliability
// features. If the charge succeeded, THE ORDER MUST BE PLACED for the customer.
// If we tried to do this inline in the mutation after the charge API call,
// there’s a chance that the database write would fail, or the email wouldn’t
// send.
//
// If those things happen in this case, we’ll send a 500 back to Stripe and it
// will retry. If it keeps failing, Stripe will notify us of the issue.
//
// Note that there is a non-zero chance that Stripe will call this twice for the
// same charge, meaning we will send 2 emails to the customer about the order.
// That’s a better choice than not sending an email at all or causing delays in
// fulfillment, so we can live with it.
export async function processChargeSucceeded(
  { emails, registryDb }: Dependencies,
  charge: charges.ICharge
) {
  const {
    metadata: {
      'webapp.name': webappName,
      'webapp.nodeEnv': webappNodeEnv,
      'order.orderKey': orderKey,
      'order.orderId': orderId,
    },
  } = charge;

  // We're going to get every Stripe event, so make sure that we’e only handling
  // our own. Also, we do the NODE_ENV check so that charges made in dev (which
  // still go to Test Stripe) don’t cause staging to send an email.
  if (
    webappName !== 'registry-certs' ||
    webappNodeEnv !== process.env.NODE_ENV
  ) {
    return;
  }

  // We slightly-hackily rely on the DB order -> JS object code from the GraphQL
  // side of things. This could be more principled.
  const order = await loadOrder(registryDb, orderId);

  if (!order) {
    throw new Error(`Order ${orderId} not found in the database`);
  }

  // By marking the payment as successful we can tell Registry they can proceed
  // with fulfillment. This method is idempotent.
  await registryDb.addPayment(
    parseInt(orderKey, 10),
    // Unix epoch seconds -> milliseconds
    new Date(charge.created * 1000),
    charge.id,
    // Per Rich, this should be the total amount charged by Stripe, including
    // their fee, not just the subtotal we receive.
    charge.amount / 100
  );

  // Sending the email is not idempotent, so we put it last and in the vast
  // majority of cases it will run once.
  await emails.sendReceiptEmail(order.contactName, order.contactEmail, {
    orderId: order.id,
    orderDate: order.date,
    shippingName: order.shippingName,
    shippingCompanyName: order.shippingCompanyName,
    shippingAddress1: order.shippingAddress1,
    shippingAddress2: order.shippingAddress2,
    shippingCity: order.shippingCity,
    shippingState: order.shippingState,
    shippingZip: order.shippingZip,
    subtotal: order.subtotal,
    serviceFee: order.serviceFee,
    total: order.total,
    fixedFee: FIXED_CC_SERVICE_FEE,
    percentageFee: PERCENTAGE_CC_SERVICE_FEE,
    serviceFeeUri: SERVICE_FEE_URI,
    // loadOrder comes from the GraphQL side of things first and foremost, so it
    // returns an async function for the certificate so that the extra DB lookup
    // (needed for certificate name) doesn’t have to happen if it’s not
    // requested by the query. Since we do need the name for the receipt, we
    // have to trigger and then resolve the promises.
    items: await Promise.all(
      order.items.map(async ({ id, quantity, cost, certificate }) => ({
        id,
        quantity,
        cost,
        name: nameFromCertificate(await certificate()),
      }))
    ),
  });
}

function nameFromCertificate(cert) {
  if (cert) {
    return `${cert.firstName} ${cert.lastName}`;
  } else {
    // typically should not happen, but we need to guard anyway
    return 'UNKNOWN CERTIFICATE';
  }
}

export async function processStripeEvent(
  deps: Dependencies,
  webhookSecret: string | undefined,
  webhookSignature: string,
  body: string
): Promise<void> {
  const event: webhooks.StripeWebhookEvent<any> = webhookSecret
    ? deps.stripe.webhooks.constructEvent(body, webhookSignature, webhookSecret)
    : JSON.parse(body);

  if (process.env['NODE_ENV'] !== 'test') {
    console.log(
      'STRIPE WEBHOOK: ',
      JSON.stringify({ ...event, data: 'REDACTED' })
    );
  }

  switch (event.type) {
    case 'charge.succeeded':
      await processChargeSucceeded(deps, event.data.object as charges.ICharge);
      break;
    default:
      break;
  }
}
