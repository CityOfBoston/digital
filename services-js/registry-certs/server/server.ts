/* eslint no-console: 0 */
import Hapi from 'hapi';
import next from 'next';
import Boom from 'boom';
import Inert from 'inert';
import fs from 'fs';
import { graphqlHapi, graphiqlHapi } from 'apollo-server-hapi';
import cleanup from 'node-cleanup';
import Stripe from 'stripe';
import { Client as PostmarkClient } from 'postmark';
import Rollbar from 'rollbar';

import {
  loggingPlugin,
  adminOkRoute,
  makeStaticAssetRoutes,
  headerKeys,
  HeaderKeysOptions,
  rollbarPlugin,
  graphqlOptionsWithRollbar,
} from '@cityofboston/hapi-common';

import { makeRoutesForNextApp, makeNextHandler } from '@cityofboston/hapi-next';

import {
  GRAPHQL_PATH_KEY,
  API_KEY_CONFIG_KEY,
  HAPI_INJECT_CONFIG_KEY,
} from '@cityofboston/next-client-common';

import decryptEnv from '@cityofboston/srv-decrypt-env';

import {
  makeRegistryDbFactory,
  makeFixtureRegistryDbFactory,
  RegistryDbFactory,
} from './services/RegistryDb';

import Emails from './services/Emails';

import { processStripeEvent } from './stripe-events';

import schema, { Context } from './graphql';
import { DatabaseConnectionOptions } from '@cityofboston/mssql-common';

type ServerArgs = {
  rollbar: Rollbar;
};

const port = parseInt(process.env.PORT || '3000', 10);

export async function makeServer({ rollbar }: ServerArgs) {
  const serverOptions = {
    port,
    ...(process.env.USE_SSL
      ? {
          tls: {
            key: fs.readFileSync('server.key'),
            cert: fs.readFileSync('server.crt'),
          },
        }
      : {}),
  };

  const server = new Hapi.Server(serverOptions);

  // We load the config ourselves so that we can modify the runtime configs
  // from here.
  const config = require('../../next.config.js');

  config.publicRuntimeConfig = {
    ...config.publicRuntimeConfig,
    [GRAPHQL_PATH_KEY]: '/graphql',
    [API_KEY_CONFIG_KEY]: process.env.WEB_API_KEY || '',
    stripePublishableKey:
      process.env.STRIPE_PUBLISHABLE_KEY || 'fake-stripe-key',
  };

  config.serverRuntimeConfig = {
    ...config.serverRuntimeConfig,
    [HAPI_INJECT_CONFIG_KEY]: server.inject.bind(server),
  };

  const app = next({
    dev: process.env.NODE_ENV !== 'production' && !process.env.USE_BUILD,
    quiet: process.env.NODE_ENV === 'test',
    config,
  });

  // These env variables are named "DATA" for historical reasons.
  const registryDbFactoryOpts: DatabaseConnectionOptions = {
    username: process.env.REGISTRY_DATA_DB_USER!,
    password: process.env.REGISTRY_DATA_DB_PASSWORD!,
    domain: process.env.REGISTRY_DATA_DB_DOMAIN,
    server: process.env.REGISTRY_DATA_DB_SERVER!,
    database: process.env.REGISTRY_DATA_DB_DATABASE!,
  };

  const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || 'fake-secret-key');
  const stripeWebhookSecret = process.env.STRIPE_WEBHOOK_SECRET;
  if (process.env.NODE_ENV === 'production' && !stripeWebhookSecret) {
    throw new Error('NEED A WEBHOOK SECRET IN PROD');
  }

  const postmarkClient = new PostmarkClient(
    process.env.POSTMARK_SERVER_API_TOKEN || 'fake-postmark-key'
  );

  const emails = new Emails(
    process.env.POSTMARK_FROM_ADDRESS || 'no-reply@boston.gov',
    postmarkClient,
    rollbar
  );

  let registryDbFactory: RegistryDbFactory;

  const startup = async () => {
    const services = await Promise.all([
      registryDbFactoryOpts.server
        ? makeRegistryDbFactory(rollbar, registryDbFactoryOpts)
        : makeFixtureRegistryDbFactory('fixtures/registry-data/smith.json'),
      app.prepare(),
    ]);

    registryDbFactory = services[0] as any;

    return async () => {
      await Promise.all([
        registryDbFactory.cleanup(),
        app.close(),
        server.stop(),
      ]);
    };
  };

  server.auth.scheme('headerKeys', headerKeys);
  server.auth.strategy('apiHeaderKeys', 'headerKeys', {
    header: 'X-API-KEY',
    keys: process.env.API_KEYS ? process.env.API_KEYS.split(',') : [],
  } as HeaderKeysOptions);

  await server.register({
    plugin: rollbarPlugin,
    options: { rollbar },
  });

  if (process.env.NODE_ENV !== 'test') {
    await server.register(loggingPlugin);
  }

  await server.register(Inert);

  await server.register({
    plugin: graphqlHapi,
    options: {
      path: '/graphql',
      // We use a function here so that all of our services are request-scoped
      // and can cache within the same query but not leak to others.
      graphqlOptions: graphqlOptionsWithRollbar(rollbar, () => ({
        schema,
        context: {
          registryDb: registryDbFactory.registryDb(),
          stripe,
          emails,
          rollbar,
        } as Context,
      })),
      route: {
        cors: true,
        auth: process.env.API_KEYS ? 'apiHeaderKeys' : false,
      },
    },
  });

  server.register({
    plugin: graphiqlHapi,
    options: {
      path: '/graphiql',
      graphiqlOptions: {
        endpointURL: '/graphql',
        passHeader: `'X-API-KEY': '${process.env.WEB_API_KEY || ''}'`,
      },
    },
  });

  server.route({
    method: 'GET',
    path: '/',
    handler: (_, h) => h.redirect(process.env.ROOT_REDIRECT_URL || '/death'),
  });

  // small hack to keep people from finding the UI in prod, since it doesnâ€™t
  // work
  if (
    process.env.NODE_ENV === 'production' &&
    process.env.BIRTH_CERTS_ENABLED !== '1'
  ) {
    server.route({
      method: 'GET',
      path: '/birth/{p*}',
      handler: () => {
        throw Boom.notFound();
      },
    });
  }

  server.route(adminOkRoute);

  // Stripe webhook handler. Used to reliably complete the order process when
  // charges succeed.
  server.route({
    method: 'POST',
    path: '/stripe',
    options: {
      payload: {
        output: 'data',
        parse: false,
      },
    },
    handler: async request => {
      try {
        await processStripeEvent(
          {
            registryDb: registryDbFactory.registryDb(),
            stripe,
            emails,
          },
          stripeWebhookSecret,
          request.headers['stripe-signature'],
          (request.payload as any).toString()
        );
        return '';
      } catch (e) {
        rollbar.error(e, request.raw.req);
        throw e;
      }
    },
  });

  server.route(makeRoutesForNextApp(app));
  server.route({
    method: 'GET',
    path: '/death/certificate/{id}',
    handler: makeNextHandler(app, '/death/certificate'),
  });

  server.route(makeStaticAssetRoutes());

  return {
    server,
    startup,
  };
}

export default async function startServer(args: ServerArgs) {
  await decryptEnv();

  const { server, startup } = await makeServer(args);

  const shutdown = await startup();
  cleanup(exitCode => {
    shutdown().then(
      () => {
        process.exit(exitCode);
      },
      err => {
        console.log('CLEAN EXIT FAILED', err);
        process.exit(-1);
      }
    );

    cleanup.uninstall();
    return false;
  });

  await server.start();

  console.log(`> Ready on http://localhost:${port}`);
}
