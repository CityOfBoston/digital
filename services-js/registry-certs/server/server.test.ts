/* eslint no-console: 0 */

import { makeServer } from './server';
import { Server } from 'hapi';
import Stripe from 'stripe';

jest.mock('stripe');

// eslint-disable-next-line no-undef
jasmine.DEFAULT_TIMEOUT_INTERVAL = 15000;

const API_KEY = 'test-api-key';
const FULFILLMENT_API_KEY = 'fulfillment-api-key';

const GRAPHQL_INTROSPECTION_PAYLOAD = {
  query: `
    {
      __schema {
        types {
          name
        }
      }
    }`,
};

let rollbar;

// Needs to be "beforeAll" not "beforeEach" so that we don’t create new ones
// with each test, since the server object holds on to the original ones.
beforeAll(() => {
  rollbar = {
    error: jest.fn(),
  };
});

describe('server', () => {
  it('can be created', async () => {
    const { server } = await makeServer({ rollbar });
    expect(server).toBeDefined();
  });
});

describe('running server', () => {
  let server: Server;
  let shutdown: () => void;

  beforeAll(async () => {
    (Stripe as any).mockImplementation(() => ({
      charges: {
        retrieve: jest.fn().mockReturnValue(Promise.resolve({})),
        refund: jest.fn().mockReturnValue(Promise.resolve({})),
      },
    }));

    const oldLog = console.log;
    console.log = () => {};

    process.env.API_KEYS = API_KEY;
    process.env.FULFILLMENT_API_KEY = FULFILLMENT_API_KEY;

    // Keeps us from colliding with other servers that might be running
    process.env.PORT = '0';

    const out = await makeServer({ rollbar });
    server = out.server;
    const startup = out.startup;

    shutdown = await startup();
    console.log = oldLog;
  });

  afterAll(async () => {
    await shutdown();
  });

  describe('graphql', () => {
    it('handles request', async () => {
      const resp = await server.inject({
        method: 'POST',
        url: '/graphql',
        headers: {
          'X-API-KEY': API_KEY,
        },
        payload: GRAPHQL_INTROSPECTION_PAYLOAD,
      });

      expect(resp.statusCode).toEqual(200);
      expect(JSON.parse(resp.payload)).toMatchSnapshot();
    });

    it('rejects wrong key', async () => {
      const resp = await server.inject({
        method: 'POST',
        url: '/graphql',
        headers: {
          'X-API-KEY': 'not-a-key',
        },
        payload: GRAPHQL_INTROSPECTION_PAYLOAD,
      });
      expect(resp.statusCode).toEqual(401);
    });

    it('rejects missing auth header', async () => {
      const resp = await server.inject({
        method: 'POST',
        url: '/graphql',
        payload: GRAPHQL_INTROSPECTION_PAYLOAD,
      });
      expect(resp.statusCode).toEqual(401);
    });

    // We test this specially because they’re a bit of a pain to do manually.
    it('handles persisted queries', async () => {
      const resp = await server.inject({
        method: 'POST',
        url: '/graphql',
        headers: {
          'X-API-KEY': FULFILLMENT_API_KEY,
        },
        payload: {
          id: 'cancel-bc-order-v1',
          variables: {
            orderId: 'test-order-id',
            transactionId: 'test-transaction-id',
          },
        },
      });

      expect(resp.statusCode).toEqual(200);

      const out = JSON.parse(resp.payload);
      expect(out).toMatchInlineSnapshot(`
        Object {
          "data": Object {
            "cancelOrder": Object {
              "error": null,
              "success": true,
            },
          },
        }
      `);
    });
  });

  describe('next', () => {
    // Put this in when we have a way for .next changes not to trigger the
    // jest watcher
    // it('loads search page', async () => {
    //   const resp = await server.inject('/death?q=smith');
    //   expect(resp.statusCode).toEqual(200);
    // });
  });
});
