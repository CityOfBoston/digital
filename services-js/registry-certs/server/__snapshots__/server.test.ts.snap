// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`running server graphql handles request 1`] = `
Object {
  "data": Object {
    "__schema": Object {
      "types": Array [
        Object {
          "name": "Query",
        },
        Object {
          "name": "DeathCertificates",
        },
        Object {
          "name": "String",
        },
        Object {
          "name": "DeathCertificate",
        },
        Object {
          "name": "Boolean",
        },
        Object {
          "name": "DeathCertificateOrder",
        },
        Object {
          "name": "Float",
        },
        Object {
          "name": "CertificateOrderItem",
        },
        Object {
          "name": "Int",
        },
        Object {
          "name": "DeathCertificateSearch",
        },
        Object {
          "name": "Mutation",
        },
        Object {
          "name": "OrderType",
        },
        Object {
          "name": "ChargeOrderResult",
        },
        Object {
          "name": "ChargeOrderError",
        },
        Object {
          "name": "ChargeOrderErrorCode",
        },
        Object {
          "name": "DeleteUploadResult",
        },
        Object {
          "name": "BirthCertificateOrderItemInput",
        },
        Object {
          "name": "OrderResult",
        },
        Object {
          "name": "OrderError",
        },
        Object {
          "name": "OrderErrorCause",
        },
        Object {
          "name": "SubmittedOrder",
        },
        Object {
          "name": "DeathCertificateOrderItemInput",
        },
        Object {
          "name": "MarriageCertificateOrderItemInput",
        },
        Object {
          "name": "__Schema",
        },
        Object {
          "name": "__Type",
        },
        Object {
          "name": "__TypeKind",
        },
        Object {
          "name": "__Field",
        },
        Object {
          "name": "__InputValue",
        },
        Object {
          "name": "__EnumValue",
        },
        Object {
          "name": "__Directive",
        },
        Object {
          "name": "__DirectiveLocation",
        },
        Object {
          "name": "Date",
        },
      ],
    },
  },
}
`;
